ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/UtxoLedgerService.kt

```diff
@@ -120,4 +121,79 @@ interface UtxoLedgerService {
         session: FlowSession,
         validator: UtxoTransactionValidator
     ): UtxoSignedTransaction
+
+    /**
+     * Sends a transaction builder to another session with all dependent back chains.
+     *
+     * @param transactionBuilder The [UtxoTransactionBuilder] to send.
+     * @param session The receiver [FlowSession].
+     */
+    @Suspendable
+    fun sendTransactionBuilder(
+        transactionBuilder: UtxoTransactionBuilder,
+        session: FlowSession
+    )
+
+    /**
+     * Receives a transaction builder from another session.
+     *
+     * @param session The [FlowSession] to receive the [UtxoTransactionBuilder] from.
+     */
+    @Suspendable
+    fun receiveTransactionBuilder(
+        session: FlowSession,
+    ): UtxoTransactionBuilder
+
+    /**
+     * Sends the differences of transaction builders to another session with all dependent back chains.
+     *
+     * @param transactionBuilder The [UtxoTransactionBuilder] to send.
+     * @param session The receiver [FlowSession].
+     * @param originalTransactionalBuilder Baseline [UtxoTransactionBuilder] to calculate the differences to be sent.
+     */
+    @Suspendable
+    fun sendTransactionBuilderDiff(
+        transactionBuilder: UtxoTransactionBuilder,
+        session: FlowSession,
+        originalTransactionalBuilder: UtxoTransactionBuilder
+    )
+
+    /**
+     * Receives a transaction builder from another session then appends its parts to an existing one.
+     *
+     * @param session The [FlowSession] to receive the [UtxoTransactionBuilder] from.
+     * @param originalTransactionalBuilder The received parts will be appended to this [UtxoTransactionBuilder].
+     */
+    @Suspendable
+    fun receiveAndUpdateTransactionBuilder(
+        session: FlowSession,
+        originalTransactionalBuilder: UtxoTransactionBuilder
+    ): UtxoTransactionBuilder
+
+    /**
+     * Sends a transaction builders to another session, then it waits for other side returning an updated builder.
```

Comment:
This one needs to ensure that the returned tx builder is strictly an addition to the one passed in - this is probably the case, given that we send diffs, but should be documented as API promise.

I wonder whether we should also have a callback to check that we agree with the diff that is returned? Or would CorDapps somehow have to check the returned tx builder? Would this modify the one passed in or return a modified copy, so we can do a diff?

Reply:
> This one needs to ensure that the returned tx builder is strictly an addition to the one passed in - this is probably the case, given that we send diffs, but should be documented as API promise.

Yes. I have updated the doc to reflect this better.
It sends only the diff on sender side, but also does not apply again anything which is already in the original builder on the receiver side.

> 
> I wonder whether we should also have a callback to check that we agree with the diff that is returned? Or would CorDapps somehow have to check the returned tx builder? Would this modify the one passed in or return a modified copy, so we can do a diff?

My original thinking was that the received builder can be used as a container only to let the cordapp cherry-pick its components into another builder, if the receiver side wants to control what it wants to accept and what not.
Then I just realized that only the `notary` is exposed on the `UtxoTransactionBuilder` interface.
So it is impossible to do any kind of diffing in the CorDapp.
If we want to give control, I would expose components on the `UtxoTransactionBuilder` and let the CorDapp decide how they want to cherry-pick.